✨(badges) add Open Badge Factory provider

jmaupetit opened the pull request:
## Purpose

Open Badge Factory is a popular Open Badge provider that we need to implement first while we are not able to generate our own badges.

## Proposal

- [x] add base provider
- [x] add OBF provider

Note that since the API access token validity is limited (10 hours), we've implemented an automatic renewal mecanism in case we must generate badges from the web server and not only from a batch job.

OBF API documentation is available at: https://openbadgefactory.com/static/doc/obf-api-v1.pdf

jmaupetit commented on src/backend/joanie/tests/test_badges_provider_obf.py:
```diff
@@ -0,0 +1,688 @@
+        )
+        assert obf.delete() is None
+
+        # An error occured while updating the badge
```
```suggestion
        # An error occurred while updating the badge
```

jmaupetit commented on src/backend/joanie/tests/test_badges_provider_obf.py:
```diff
@@ -0,0 +1,688 @@
+            is None
+        )
+
+        # An error occured while updating the badge
```
```suggestion
        # An error occurred while updating the badge
```

jbpenrath commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,468 @@
+"""OpenBadgeFactory provider."""
```
For your information, during your holidays, we refactor tests architecture to centralize them within the same folder.
You'll have to move this test suite accordingly.
https://github.com/openfun/joanie/pull/128

jmaupetit replied:
<img src="https://media0.giphy.com/media/xakXSnCpsqZWM/giphy.gif"/>

jbpenrath commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,468 @@
+            }
+        )
+
+    @cached_property
```
Are you sure `@cached_property` serves your purpose here as the cached value relies on the class instance, so a request to OBF will be send for each new `OBFAPIClient` instance.

jmaupetit replied:
We can bet that a script or management command execution will rely on a single API client instance for a limited amount of time (the script execution time), hence, caching this token avoids renewing it at each API request. What alternative approach would you suggest?

jmaupetit replied:
After chatting with @sampaccoud, I now understand that you expect a more reliable cache to rely on in HTTP requests processing context. Hence, I refactored this method as a LRU cached static method. Hope this was what you expected.

jbpenrath replied:
This is exactly what I expected 👌

jbpenrath commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,468 @@
+
+    @classmethod
+    def iter_json(cls, response: requests.Response) -> Iterable:
+        """Interater over JSON lines serialization in API responses.
```
```suggestion
        """Iterate over JSON lines serialization in API responses.
```

sampaccoud commented on src/backend/joanie/badges/exceptions.py:
```diff
@@ -0,0 +1,9 @@
+
+
+class BadgeProviderError(Exception):
+    """Generic error raised when a Badge provider encounter an exception."""
```
```suggestion
    """Generic error raised when a Badge provider encounters an exception."""
```

sampaccoud commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,468 @@
+
+        if "access_token" not in response.json():
+            msg = "Cannot get an access token from the OBF server with provided credentials"
+            logger.error(msg)
+            logger.debug("Request status code was: %s", response.status_code)
+            logger.debug("Request URL was: %s", response.request.url)
+            logger.debug("Request body was: %s", response.request.body)
+            logger.debug("Response text was: %s", response.text)
```
I would say this would create 5 separate events in Sentry for each execution and raising the error in the next line will create one more. Since you know you're going to raise, why not include this info in the error raised?

sampaccoud commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,468 @@
+            },
+        )
+
+        if "access_token" not in response.json():
```
I would say there are cases when response.json() will fail so you need to first test if the request was successful?

sampaccoud commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,468 @@
+        try:
+            json_ = response.json()
+            if isinstance(json_, list):
+                for item in json_:
+                    yield item
```
```suggestion
                yield from json_
```

Maybe you can use a try/except pattern?
```
try:
    yield from json_
except ...:
    yield json_

jmaupetit replied:
<img src="https://media0.giphy.com/media/Od0QRnzwRBYmDU3eEO/giphy.gif"/>

jmaupetit replied:
In fact, as a dict is an iterator, I opt for the following:

```python
json_response = response.json()
if isinstance(json_response, list):
    yield from json_response
else:
    yield json_response
```

sampaccoud commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,468 @@
+                response.status_code,
+            )
+            if self.raise_for_status:
+                raise error
```
should be either logger.error or raise but not both?

jmaupetit replied:
True. So I should just raise for status if asked to.

jmaupetit commented:
@sampaccoud I've addressed all your comments and improved my authentication cache implementation following @jbpenrath comment and our discussion. I've pushed changes in fixup commits so that it's easier to (re-)review.

jbpenrath commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,472 @@
+            raise BadgeProviderError(f"Cannot issue badge with ID: {badge.id}")
+
+        event_url = response.headers.get("Location")
+        logging.info(
```
```suggestion
        logger.info(
```

jbpenrath commented on src/backend/joanie/badges/providers/obf.py:
```diff
@@ -0,0 +1,472 @@
+            len(issue.recipient),
+            badge.id,
+        )
+        logging.info("Issued badges event URL: %s", event_url)
```
```suggestion
        logger.info("Issued badges event URL: %s", event_url)
```

jbpenrath reviewed (approved):
LGTM ✅ 🎉
